### ISSUE TITLE
Types issues

### ISSUE THREAD

**Issue by shovelmn12** [rxRust/rxRust#253]
Hello

I am working on a small project to create a backend framework to handle events (microservices)
I am trying to implement the pipe to handle the events but I keep getting errors when calling `subscribe()`

Generally I dont get why is the implementation working this way.
1. Why dont extend `Stream`?
2. Why not return a unified definition of `Observable`? Each OP returns a different defenition therefore creating a mess when trying to read errors (see below)

For reference check [`rxdart`](https://github.com/ReactiveX/rxdart/tree/master). They started with their own definition then went back to create operation extension on `Stream` from dart to solve all those issues, also each Op returns `Stream` and not `MapOp`, `FilterOp` etc ...

The code:

```rust
#[derive(Clone)]
pub struct EventSource<Context, Event, Error>
where
    Context: IntoBody + Sized + Send + Sync + 'static,
    Event: IntoBody + Sized + Send + Sync + 'static,
    Error: IntoBody + Sized + Send + Sync + 'static,
{
    pub tx: Subject<'static, (Context, Event), Error>,
    pub rx: Subject<'static, (Context, Event), Error>,
}

pub type EventHandler<Context, Event, Error> =
    fn((Context, Event)) -> Pin<Box<dyn Future<Output = Result<(), Error>>>>;
pub type CreateSource<Context, Event, Error> =
    fn() -> Pin<Box<dyn Future<Output = Result<EventSource<Context, Event, Error>, Error>>>>;

pub fn start_service<Context, Event, Error>(
    source: CreateSource<Context, Event, Error>,
    handler: EventHandler<Context, Event, Error>,
) -> Box<dyn Subscription>
where
    Context: IntoBody + Sized + Sync + Send,
    Event: IntoBody + Sized + Sync + Send,
    Error: IntoBody + Sized + Sync + Send,
{
    let threads_scheduler = futures::executor::ThreadPool::new().unwrap();

    Box::new(
        from_future_result(source(), threads_scheduler.clone())
            .flat_map_threads(|source| source.rx)
            .flat_map_threads(|event| from_future_result(handler(event), threads_scheduler.clone()))
            .on_error(|error| {})
            .subscribe(|result| {}), // issue is here
    )
}
```

Full error:

<details>

```

    the method `subscribe` exists for struct `OnErrorOp<MergeAllOpThreads<MapOp<MergeAllOpThreads<MapOp<..., ..., ...>, ...>, ..., ...>, ...>, ..., ...>`, but its trait bounds were not satisfied [E0599] Note: the full type name has been written to '/ Users/ shovelmn12/ Projects/ poc_server/ target/ debug/ deps/ poc_server-b47a2d7829a631fd. long-type-1706698521457854592.txt' Note: consider using `--verbose` to print the full type name to the console Note: the following trait bounds were not satisfied: `rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::Observable<_, std::convert::Infallible, rxrust::observable::ObserverItem<_>>` which is required by `rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::ObservableItem<_, _>` `&rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::Observable<_, std::convert::Infallible, rxrust::observable::ObserverItem<_>>` which is required by `&rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::ObservableItem<_, _>` `&mut rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::Observable<_, std::convert::Infallible, rxrust::observable::ObserverItem<_>>` which is required by `&mut rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::ObservableItem<_, _>`
    the method `subscribe` exists for struct `OnErrorOp<MergeAllOpThreads<MapOp<MergeAllOpThreads<MapOp<..., ..., ...>, ...>, ..., ...>, ...>, ..., ...>`, but its trait bounds were not satisfied [E0599] Note: the full type name has been written to '/ Users/ shovelmn12/ Projects/ poc_server/ target/ debug/ deps/ poc_server-deb9ac863f34f807.long-type-7867709162962570806.txt' Note: consider using `--verbose` to print the full type name to the console Note: the following trait bounds were not satisfied: `rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::Observable<_, std::convert::Infallible, rxrust::observable::ObserverItem<_>>` which is required by `rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::ObservableItem<_, _>` `&rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::Observable<_, std::convert::Infallible, rxrust::observable::ObserverItem<_>>` which is required by `&rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::ObservableItem<_, _>` `&mut rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::Observable<_, std::convert::Infallible, rxrust::observable::ObserverItem<_>>` which is required by `&mut rxrust::ops::on_error::OnErrorOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::ops::merge_all::MergeAllOpThreads<rxrust::ops::map::MapOp<rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<EventSource<Context, Event, Error>, Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>, {closure@src/ main. rs:54:31: 54:39}, EventSource<Context, Event, Error>>, rxrust::subject::Subject<'static, (Context, Event), Error>>, {closure@src/ main. rs:55:31: 55:38}, (Context, Event)>, rxrust::observable::from_future::FutureResultObservable<std::pin::Pin<std::boxed::Box<dyn futures::Future<Output = std::result::Result<(), Error>>>>, rxrust::scheduler::FuturesThreadPoolScheduler>>, {closure@src/ main. rs:56:23: 56:30}, Error>: rxrust::observable::ObservableItem<_, _>`


```

</details>

FYI: not a rust expert, just honestly trying to understand :)

**Comment by shovelmn12** [comment]
@M-Adoo Thanks for your response

Yes it fixed my example but unfortunately I still can't make my code work :( I'd love if you are able to check the code and let me know what's wrong there, I made another simple example (I think this is more correct than the other one):

```
    from_future_result(
        ready(Ok(Subject::<(Context, Event), Error>::default())), // the subject is coming as input param for producing events
        threads_scheduler,
    )
    .flat_map_threads(|s| s) // flat map that source
    .flat_map_threads(|s| from_future_result(ready(Ok(())), threads_scheduler.clone())) // call to handler of our events, mocked it with an Ok result, shuold work when returning Err too 
    .on_error(|error| {}) // handle all errors here
    .subscribe(|e| {}); // done processing the event, Still getting error here
```

I'd love to know whats the issue, it's hard to understand the error as it's so explicit!

1. Ok, got you! FYI: To my knowladge (might be wrong) `Stream` is a poll based but it uses some kernel level push based API, something like the kernel will let us know once the poll is ready. Is it not?

2. Ok great :) coming from rxdart I love the simplicity, I guess rust can do the same something in the lines of each operator is a trait, then the trait is implemented on `Observable` and returns `Observable` too. Then you got a modular way of writing all the operations + all operations would return `Observable` :) haha just sharing an idea...

```
trait Observable<Item> {}

trait MapOp {}

impl <T> MapOp for Observable<T> {
  fn map(...) -> Observable<T>
}
```

Last question, why there is always `op` and `op_threads` ?? if I use threads I'd always need to use `op_threads`? why not make all operations thread safe so there is no question what you can or cant use ...

**Comment by M-Adoo** [comment]
1. The runtime controls how to poll the `Future`.
2. A local thread version exists that does not require the closure to satisfy `Send+Sync`, for example, you can use `std::rc::Rc`.

To simplify the types, I propose updating the trait as follows:

```rust
trait Observable<Item, Err> { 
  fn map<U>(self, f: impl Fn(Item) -> U) -> impl Observable<U, Err>;
}
```

I haven't tested your code, but I believe the issue may lie here:

`flat_map_threads(|s| from_future_result(ready(Ok(())), threads_scheduler.clone()))`

You are attempting to capture `threads_scheduler` and make it thread-safe with a `static` lifetime, which is not possible.

You should clone the `threads_scheduler` outside the closure and move it into the capture like this:

```rust
let threads_scheduler = threads_scheduler.clone();
xxx
    .flat_map_threads(move |s| from_future_result(ready(Ok(())), threads_scheduler.clone())
```

**Comment by M-Adoo** [comment]
> @M-Adoo BTW I'd love to contribute ...

Welcome!

**Comment by shovelmn12** [comment]
@M-Adoo thanks for your response.

If you check all the examples I added that DOES NOT WORK, none of them actually uses `threads_scheduler` inside `flat_map_threads`. Example:

```
    from_future(
        ready(Subject::<String, &str>::default()),
        threads_scheduler.clone(),
    )
    .flat_map(|source| source.on_error(|e| {}))
    .subscribe(|e| {}) // error here
```

**Comment by M-Adoo** [comment]
If you are utilizing `threads_scheduler`, you should use `flat_map_threads` due to the requirement for thread safety downstream.

In your previous demo, another issue was the need to use `SubjectThreads` instead of `Subject`. You can address this by trying the following:

```rust
let threads_scheduler = futures::executor::ThreadPool::new().unwrap();
use std::future::ready;

from_future(
    // fix: use `SubjectThreads` instead of `Subject`
    ready(SubjectThreads::<String, &str>::default()),
    threads_scheduler.clone(),
)
// fix: use `flat_map_threads` instead of `flat_map`
.flat_map_threads(|source| source.on_error(|e| {}))
.subscribe(|e| {});
```

**Comment by shovelmn12** [comment]
@M-Adoo Ok so I managed to get it working at last, thanks to your solution + another isse was that there was something wrong with my defenitions:

```
pub type EventHandler<Context, Event, Error> =
    fn((Context, Event)) -> Pin<Box<dyn Future<Output = Result<(), Error>>>>;
pub type CreateSource<Context, Event, Error> =
    fn() -> Pin<Box<dyn Future<Output = Result<EventSource<Context, Event, Error>, Error>>>>;
```

In stead of using `Pin<Box<...>>` I used `BoxFuture<...>` it solved the other errors I had :)

FYI: if you know what would you like to implement so we can streamline `Observable` let me know
